File: compat/suites/python-sdk/groups/sns.py
"""
groups/sns.py — SNS compatibility test implementations for the Python suite.
"""

from __future__ import annotations
import json
import time
from lib.harness import TestContext
from lib.clients import make_clients


def _sns(ctx: TestContext):
    return make_clients(ctx.endpoint, ctx.region).sns


def _sqs(ctx: TestContext):
    return make_clients(ctx.endpoint, ctx.region).sqs


# ── sns-topics ────────────────────────────────────────────────────────────────

def setup_sns_topics(ctx: TestContext) -> None:
    sns = _sns(ctx)
    resp = sns.create_topic(Name=f"{ctx.run_id}-sns-t")
    ctx["sns_topic_arn"] = resp["TopicArn"]


def teardown_sns_topics(ctx: TestContext) -> None:
    arn = ctx.get("sns_topic_arn")
    if arn:
        try:
            _sns(ctx).delete_topic(TopicArn=arn)
        except Exception:
            pass


def CreateTopic(ctx: TestContext) -> None:
    sns = _sns(ctx)
    resp = sns.create_topic(Name=f"{ctx.run_id}-sns-create")
    arn = resp.get("TopicArn")
    if not arn:
        raise AssertionError("CreateTopic: missing TopicArn")
    sns.delete_topic(TopicArn=arn)


def ListTopics(ctx: TestContext) -> None:
    sns = _sns(ctx)
    arn = ctx["sns_topic_arn"]
    paginator = sns.get_paginator("list_topics")
    arns = []
    for page in paginator.paginate():
        arns.extend(t["TopicArn"] for t in page.get("Topics", []))
    if arn not in arns:
        raise AssertionError(f"ListTopics: {arn!r} not found in topic list")


def GetTopicAttributes(ctx: TestContext) -> None:
    sns = _sns(ctx)
    arn = ctx["sns_topic_arn"]
    resp = sns.get_topic_attributes(TopicArn=arn)
    if "TopicArn" not in resp.get("Attributes", {}):
        raise AssertionError(f"GetTopicAttributes: missing TopicArn in attributes {resp.get('Attributes')}")


def SetTopicAttributes(ctx: TestContext) -> None:
    sns = _sns(ctx)
    arn = ctx["sns_topic_arn"]
    sns.set_topic_attributes(TopicArn=arn, AttributeName="DisplayName", AttributeValue="CompatTest")
    resp = sns.get_topic_attributes(TopicArn=arn)
    if resp["Attributes"].get("DisplayName") != "CompatTest":
        raise AssertionError(f"SetTopicAttributes: DisplayName not updated; got {resp['Attributes'].get('DisplayName')!r}")


def DeleteTopic(ctx: TestContext) -> None:
    sns = _sns(ctx)
    resp = sns.create_topic(Name=f"{ctx.run_id}-sns-del")
    arn = resp["TopicArn"]
    sns.delete_topic(TopicArn=arn)
    list_resp = sns.list_topics()
    arns = [t["TopicArn"] for t in list_resp.get("Topics", [])]
    if arn in arns:
        raise AssertionError(f"DeleteTopic: {arn!r} still listed after deletion")


# ── sns-publish ───────────────────────────────────────────────────────────────

def setup_sns_publish(ctx: TestContext) -> None:
    sns = _sns(ctx)
    resp = sns.create_topic(Name=f"{ctx.run_id}-sns-pub")
    ctx["sns_pub_arn"] = resp["TopicArn"]


def teardown_sns_publish(ctx: TestContext) -> None:
    arn = ctx.get("sns_pub_arn")
    if arn:
        try:
            _sns(ctx).delete_topic(TopicArn=arn)
        except Exception:
            pass


def Publish(ctx: TestContext) -> None:
    sns = _sns(ctx)
    arn = ctx["sns_pub_arn"]
    resp = sns.publish(TopicArn=arn, Message="hello from python")
    if not resp.get("MessageId"):
        raise AssertionError("Publish: missing MessageId")


def PublishWithAttributes(ctx: TestContext) -> None:
    sns = _sns(ctx)
    arn = ctx["sns_pub_arn"]
    resp = sns.publish(
        TopicArn=arn,
        Message="message with attrs",
        MessageAttributes={
            "source": {"DataType": "String", "StringValue": "compat-test"}
        },
    )
    if not resp.get("MessageId"):
        raise AssertionError("PublishWithAttributes: missing MessageId")


def PublishBatch(ctx: TestContext) -> None:
    sns = _sns(ctx)
    arn = ctx["sns_pub_arn"]
    resp = sns.publish_batch(
        TopicArn=arn,
        PublishBatchRequestEntries=[
            {"Id": "1", "Message": "batch msg 1"},
            {"Id": "2", "Message": "batch msg 2"},
        ],
    )
    if len(resp.get("Successful", [])) < 2:
        raise AssertionError(f"PublishBatch: expected 2 successful, got {resp}")


# ── sns-subscriptions ─────────────────────────────────────────────────────────

def setup_sns_subscriptions(ctx: TestContext) -> None:
    sns = _sns(ctx)
    sqs = _sqs(ctx)
    topic_resp = sns.create_topic(Name=f"{ctx.run_id}-sns-sub")
    ctx["sns_sub_topic_arn"] = topic_resp["TopicArn"]
    q_resp = sqs.create_queue(QueueName=f"{ctx.run_id}-sns-sub-q")
    ctx["sns_sub_queue_url"] = q_resp["QueueUrl"]
    q_attrs = sqs.get_queue_attributes(
        QueueUrl=q_resp["QueueUrl"], AttributeNames=["QueueArn"]
    )
    ctx["sns_sub_queue_arn"] = q_attrs["Attributes"]["QueueArn"]


def teardown_sns_subscriptions(ctx: TestContext) -> None:
    sns = _sns(ctx)
    sqs = _sqs(ctx)
    sub_arn = ctx.get("sns_sub_arn")
    if sub_arn:
        try:
            sns.unsubscribe(SubscriptionArn=sub_arn)
        except Exception:
            pass
    topic_arn = ctx.get("sns_sub_topic_arn")
    if topic_arn:
        try:
            sns.delete_topic(TopicArn=topic_arn)
        except Exception:
            pass
    queue_url = ctx.get("sns_sub_queue_url")
    if queue_url:
        try:
            sqs.delete_queue(QueueUrl=queue_url)
        except Exception:
            pass


def SubscribeSQS(ctx: TestContext) -> None:
    sns = _sns(ctx)
    topic_arn = ctx["sns_sub_topic_arn"]
    queue_arn = ctx["sns_sub_queue_arn"]
    resp = sns.subscribe(
        TopicArn=topic_arn,
        Protocol="sqs",
        Endpoint=queue_arn,
        Attributes={"RawMessageDelivery": "true"},
    )
    if not resp.get("SubscriptionArn"):
        raise AssertionError("SubscribeSQS: missing SubscriptionArn")
    ctx["sns_sub_arn"] = resp["SubscriptionArn"]


def ListSubscriptionsByTopic(ctx: TestContext) -> None:
    sns = _sns(ctx)
    topic_arn = ctx["sns_sub_topic_arn"]
    resp = sns.list_subscriptions_by_topic(TopicArn=topic_arn)
    subs = resp.get("Subscriptions", [])
    if not subs:
        raise AssertionError("ListSubscriptionsByTopic: no subscriptions returned")


def GetSubscriptionAttributes(ctx: TestContext) -> None:
    sns = _sns(ctx)
    sub_arn = ctx.get("sns_sub_arn")
    if not sub_arn:
        raise AssertionError("GetSubscriptionAttributes: no subscription ARN")
    resp = sns.get_subscription_attributes(SubscriptionArn=sub_arn)
    if "SubscriptionArn" not in resp.get("Attributes", {}):
        raise AssertionError(f"GetSubscriptionAttributes: missing SubscriptionArn; got {resp.get('Attributes')}")


def PublishDeliveredToSQS(ctx: TestContext) -> None:
    sns = _sns(ctx)
    sqs = _sqs(ctx)
    topic_arn = ctx["sns_sub_topic_arn"]
    queue_url = ctx["sns_sub_queue_url"]
    sns.publish(TopicArn=topic_arn, Message="deliver me")
    # Poll until message arrives (or timeout)
    deadline = time.monotonic() + 5
    while time.monotonic() < deadline:
        resp = sqs.receive_message(QueueUrl=queue_url, MaxNumberOfMessages=1, WaitTimeSeconds=0)
        msgs = resp.get("Messages", [])
        if msgs:
            sqs.delete_message(QueueUrl=queue_url, ReceiptHandle=msgs[0]["ReceiptHandle"])
            return
        time.sleep(0.1)
    raise AssertionError("PublishDeliveredToSQS: message not delivered within 5s")


def SetSubscriptionAttributes(ctx: TestContext) -> None:
    sns = _sns(ctx)
    sub_arn = ctx.get("sns_sub_arn")
    if not sub_arn:
        raise AssertionError("SetSubscriptionAttributes: no subscription ARN")
    sns.set_subscription_attributes(
        SubscriptionArn=sub_arn,
        AttributeName="RawMessageDelivery",
        AttributeValue="false",
    )
    attrs = sns.get_subscription_attributes(SubscriptionArn=sub_arn)
    assert attrs["Attributes"].get("RawMessageDelivery") == "false", "SetSubscriptionAttributes: value not set"


def Unsubscribe(ctx: TestContext) -> None:
    sns = _sns(ctx)
    sub_arn = ctx.get("sns_sub_arn")
    if not sub_arn:
        raise AssertionError("Unsubscribe: no subscription ARN")
    topic_arn = ctx.get("sns_topic_arn")
    sns.unsubscribe(SubscriptionArn=sub_arn)
    ctx["sns_sub_arn"] = None
    if topic_arn:
        subs = sns.list_subscriptions_by_topic(TopicArn=topic_arn)
        arns = [s["SubscriptionArn"] for s in subs.get("Subscriptions", [])]
        assert sub_arn not in arns, f"Unsubscribe: subscription {sub_arn} still present"


# ── ImplMap ───────────────────────────────────────────────────────────────────

IMPLS = {
    "CreateTopic": CreateTopic,
    "ListTopics": ListTopics,
    "GetTopicAttributes": GetTopicAttributes,
    "SetTopicAttributes": SetTopicAttributes,
    "DeleteTopic": DeleteTopic,
    "Publish": Publish,
    "PublishWithAttributes": PublishWithAttributes,
    "PublishBatch": PublishBatch,
    "SubscribeSQS": SubscribeSQS,
    "ListSubscriptionsByTopic": ListSubscriptionsByTopic,
    "GetSubscriptionAttributes": GetSubscriptionAttributes,
    "PublishDeliveredToSQS": PublishDeliveredToSQS,
    "SetSubscriptionAttributes": SetSubscriptionAttributes,
    "Unsubscribe": Unsubscribe,
}

SETUP = {
    "sns-topics": setup_sns_topics,
    "sns-publish": setup_sns_publish,
    "sns-subscriptions": setup_sns_subscriptions,
}

TEARDOWN = {
    "sns-topics": teardown_sns_topics,
    "sns-publish": teardown_sns_publish,
    "sns-subscriptions": teardown_sns_subscriptions,
}

File: compat/suites/go-sdk/internal/groups/sns.go
package groups

import (
	"context"
	"fmt"
	"time"

	"github.com/Neaox/overcast-compat-go-sdk/internal/clients"
	"github.com/Neaox/overcast-compat-go-sdk/internal/harness"
	"github.com/aws/aws-sdk-go-v2/aws"
	"github.com/aws/aws-sdk-go-v2/service/sns"
	snstypes "github.com/aws/aws-sdk-go-v2/service/sns/types"
	"github.com/aws/aws-sdk-go-v2/service/sqs"
	sqstypes "github.com/aws/aws-sdk-go-v2/service/sqs/types"
)

func SNS(c *clients.Clients) ServiceGroup {
	g := &snsGroup{c: c}
	return ServiceGroup{
		Impls: map[string]harness.TestFn{
			"CreateTopic":               g.CreateTopic,
			"GetTopicAttributes":        g.GetTopicAttributes,
			"SetTopicAttributes":        g.SetTopicAttributes,
			"ListTopics":                g.ListTopics,
			"DeleteTopic":               g.DeleteTopic,
			"Publish":                   g.Publish,
			"PublishBatch":              g.PublishBatch,
			"PublishWithAttributes":     g.PublishWithAttributes,
			"PublishDeliveredToSQS":     g.PublishDeliveredToSQS,
			"SubscribeSQS":              g.SubscribeSQS,
			"ListSubscriptionsByTopic":  g.ListSubscriptionsByTopic,
			"GetSubscriptionAttributes": g.GetSubscriptionAttributes,
			"SetSubscriptionAttributes": g.SetSubscriptionAttributes,
			"Unsubscribe":               g.Unsubscribe,
		},
		Setup: map[string]func(context.Context, *harness.TestContext) error{
			"sns-topics":        g.setupTopics,
			"sns-publish":       g.setupPublish,
			"sns-subscriptions": g.setupSubscriptions,
		},
		Teardown: map[string]func(context.Context, *harness.TestContext) error{
			"sns-topics":        g.teardownTopics,
			"sns-publish":       g.teardownPublish,
			"sns-subscriptions": g.teardownSubscriptions,
		},
	}
}

type snsGroup struct{ c *clients.Clients }

func (g *snsGroup) sns() *sns.Client { return g.c.SNS() }
func (g *snsGroup) sqs() *sqs.Client { return g.c.SQS() }

// ── sns-topics ────────────────────────────────────────────────────────────────

func (g *snsGroup) setupTopics(ctx context.Context, t *harness.TestContext) error {
	name := fmt.Sprintf("%s-sns", t.RunID)
	resp, err := g.sns().CreateTopic(ctx, &sns.CreateTopicInput{Name: aws.String(name)})
	if err != nil {
		return err
	}
	t.Set("sns_topic_arn", aws.ToString(resp.TopicArn))
	t.Set("sns_topic_name", name)
	return nil
}

func (g *snsGroup) teardownTopics(ctx context.Context, t *harness.TestContext) error {
	if arn := t.GetString("sns_topic_arn"); arn != "" {
		g.sns().DeleteTopic(ctx, &sns.DeleteTopicInput{TopicArn: aws.String(arn)}) //nolint:errcheck
	}
	return nil
}

func (g *snsGroup) CreateTopic(ctx context.Context, t *harness.TestContext) error {
	name := fmt.Sprintf("%s-snscreate", t.RunID)
	resp, err := g.sns().CreateTopic(ctx, &sns.CreateTopicInput{Name: aws.String(name)})
	if err != nil {
		return err
	}
	defer g.sns().DeleteTopic(ctx, &sns.DeleteTopicInput{TopicArn: resp.TopicArn}) //nolint:errcheck
	// Verify topic appears in ListTopics
	list, err := g.sns().ListTopics(ctx, &sns.ListTopicsInput{})
	if err != nil {
		return fmt.Errorf("CreateTopic: ListTopics verify failed: %w", err)
	}
	for _, topic := range list.Topics {
		if aws.ToString(topic.TopicArn) == aws.ToString(resp.TopicArn) {
			return nil
		}
	}
	return fmt.Errorf("CreateTopic: topic not found in ListTopics")
}

func (g *snsGroup) GetTopicAttributes(ctx context.Context, t *harness.TestContext) error {
	arn := t.GetString("sns_topic_arn")
	resp, err := g.sns().GetTopicAttributes(ctx, &sns.GetTopicAttributesInput{TopicArn: aws.String(arn)})
	if err != nil {
		return err
	}
	if resp.Attributes["TopicArn"] != arn {
		return fmt.Errorf("GetTopicAttributes: ARN mismatch")
	}
	return nil
}

func (g *snsGroup) SetTopicAttributes(ctx context.Context, t *harness.TestContext) error {
	arn := t.GetString("sns_topic_arn")
	_, err := g.sns().SetTopicAttributes(ctx, &sns.SetTopicAttributesInput{
		TopicArn:       aws.String(arn),
		AttributeName:  aws.String("DisplayName"),
		AttributeValue: aws.String("CompatSuite"),
	})
	if err != nil {
		return err
	}
	// Verify the attribute was updated
	resp, err := g.sns().GetTopicAttributes(ctx, &sns.GetTopicAttributesInput{TopicArn: aws.String(arn)})
	if err != nil {
		return fmt.Errorf("SetTopicAttributes: GetTopicAttributes failed: %w", err)
	}
	if resp.Attributes["DisplayName"] != "CompatSuite" {
		return fmt.Errorf("SetTopicAttributes: expected DisplayName=CompatSuite, got %q", resp.Attributes["DisplayName"])
	}
	return nil
}

func (g *snsGroup) ListTopics(ctx context.Context, t *harness.TestContext) error {
	arn := t.GetString("sns_topic_arn")
	resp, err := g.sns().ListTopics(ctx, &sns.ListTopicsInput{})
	if err != nil {
		return err
	}
	for _, tp := range resp.Topics {
		if aws.ToString(tp.TopicArn) == arn {
			return nil
		}
	}
	return fmt.Errorf("ListTopics: %q not found", arn)
}

func (g *snsGroup) DeleteTopic(ctx context.Context, t *harness.TestContext) error {
	name := fmt.Sprintf("%s-snsdel", t.RunID)
	resp, err := g.sns().CreateTopic(ctx, &sns.CreateTopicInput{Name: aws.String(name)})
	if err != nil {
		return err
	}
	if _, err = g.sns().DeleteTopic(ctx, &sns.DeleteTopicInput{TopicArn: resp.TopicArn}); err != nil {
		return err
	}
	// Verify the topic is gone
	listResp, err := g.sns().ListTopics(ctx, &sns.ListTopicsInput{})
	if err != nil {
		return fmt.Errorf("DeleteTopic: ListTopics failed: %w", err)
	}
	for _, tp := range listResp.Topics {
		if aws.ToString(tp.TopicArn) == aws.ToString(resp.TopicArn) {
			return fmt.Errorf("DeleteTopic: topic %q still present after delete", aws.ToString(resp.TopicArn))
		}
	}
	return nil
}

// ── sns-publish ───────────────────────────────────────────────────────────────

func (g *snsGroup) setupPublish(ctx context.Context, t *harness.TestContext) error {
	name := fmt.Sprintf("%s-snspub", t.RunID)
	resp, err := g.sns().CreateTopic(ctx, &sns.CreateTopicInput{Name: aws.String(name)})
	if err != nil {
		return err
	}
	t.Set("sns_pub_topic_arn", aws.ToString(resp.TopicArn))

	// Create SQS queue and subscribe
	qname := fmt.Sprintf("%s-snssqs", t.RunID)
	qresp, err := g.sqs().CreateQueue(ctx, &sqs.CreateQueueInput{QueueName: aws.String(qname)})
	if err != nil {
		return err
	}
	qurl := aws.ToString(qresp.QueueUrl)
	t.Set("sns_sqs_url", qurl)

	attr, err := g.sqs().GetQueueAttributes(ctx, &sqs.GetQueueAttributesInput{
		QueueUrl: qresp.QueueUrl, AttributeNames: []sqstypes.QueueAttributeName{"QueueArn"},
	})
	if err != nil {
		return err
	}
	qarn := attr.Attributes["QueueArn"]

	subResp, err := g.sns().Subscribe(ctx, &sns.SubscribeInput{
		TopicArn: resp.TopicArn,
		Protocol: aws.String("sqs"),
		Endpoint: aws.String(qarn),
	})
	if err != nil {
		return err
	}
	t.Set("sns_sub_arn", aws.ToString(subResp.SubscriptionArn))
	return nil
}

func (g *snsGroup) teardownPublish(ctx context.Context, t *harness.TestContext) error {
	if arn := t.GetString("sns_sub_arn"); arn != "" {
		g.sns().Unsubscribe(ctx, &sns.UnsubscribeInput{SubscriptionArn: aws.String(arn)}) //nolint:errcheck
	}
	if arn := t.GetString("sns_pub_topic_arn"); arn != "" {
		g.sns().DeleteTopic(ctx, &sns.DeleteTopicInput{TopicArn: aws.String(arn)}) //nolint:errcheck
	}
	if url := t.GetString("sns_sqs_url"); url != "" {
		g.sqs().DeleteQueue(ctx, &sqs.DeleteQueueInput{QueueUrl: aws.String(url)}) //nolint:errcheck
	}
	return nil
}

func (g *snsGroup) Publish(ctx context.Context, t *harness.TestContext) error {
	arn := t.GetString("sns_pub_topic_arn")
	resp, err := g.sns().Publish(ctx, &sns.PublishInput{
		TopicArn: aws.String(arn),
		Message:  aws.String("hello-from-go-sdk"),
	})
	if err != nil {
		return err
	}
	if aws.ToString(resp.MessageId) == "" {
		return fmt.Errorf("Publish: empty MessageId")
	}
	return nil
}

func (g *snsGroup) PublishBatch(ctx context.Context, t *harness.TestContext) error {
	arn := t.GetString("sns_pub_topic_arn")
	entries := make([]snstypes.PublishBatchRequestEntry, 5)
	for i := range entries {
		entries[i] = snstypes.PublishBatchRequestEntry{
			Id: aws.String(fmt.Sprintf("m%d", i)), Message: aws.String(fmt.Sprintf("batch-%d", i)),
		}
	}
	resp, err := g.sns().PublishBatch(ctx, &sns.PublishBatchInput{
		TopicArn: aws.String(arn), PublishBatchRequestEntries: entries,
	})
	if err != nil {
		return err
	}
	if len(resp.Failed) > 0 {
		return fmt.Errorf("PublishBatch: %d failed", len(resp.Failed))
	}
	return nil
}

func (g *snsGroup) PublishWithAttributes(ctx context.Context, t *harness.TestContext) error {
	arn := t.GetString("sns_pub_topic_arn")
	resp, err := g.sns().Publish(ctx, &sns.PublishInput{
		TopicArn: aws.String(arn),
		Message:  aws.String("hello-with-attrs"),
		MessageAttributes: map[string]snstypes.MessageAttributeValue{
			"env": {DataType: aws.String("String"), StringValue: aws.String("test")},
		},
	})
	if err != nil {
		return err
	}
	if aws.ToString(resp.MessageId) == "" {
		return fmt.Errorf("PublishWithAttributes: empty MessageId")
	}
	return nil
}

func (g *snsGroup) PublishToSQS(ctx context.Context, t *harness.TestContext) error {
	return g.Publish(ctx, t) // reuses Publish
}

func (g *snsGroup) PublishDeliveredToSQS(ctx context.Context, t *harness.TestContext) error {
	arn := t.GetString("sns_sub_topic_arn")
	qurl := t.GetString("sns_sub_q_url")
	if _, err := g.sns().Publish(ctx, &sns.PublishInput{
		TopicArn: aws.String(arn), Message: aws.String("delivery-test"),
	}); err != nil {
		return err
	}
	deadline := time.Now().Add(5 * time.Second)
	for time.Now().Before(deadline) {
		recv, _ := g.sqs().ReceiveMessage(ctx, &sqs.ReceiveMessageInput{
			QueueUrl: aws.String(qurl), MaxNumberOfMessages: 10, WaitTimeSeconds: 1,
		})
		if recv != nil && len(recv.Messages) > 0 {
			for _, m := range recv.Messages {
				g.sqs().DeleteMessage(ctx, &sqs.DeleteMessageInput{ //nolint:errcheck
					QueueUrl: aws.String(qurl), ReceiptHandle: m.ReceiptHandle,
				})
			}
			return nil
		}
	}
	return fmt.Errorf("PublishDeliveredToSQS: message not delivered within 5s")
}

// ── sns-subscriptions ─────────────────────────────────────────────────────────

func (g *snsGroup) setupSubscriptions(ctx context.Context, t *harness.TestContext) error {
	name := fmt.Sprintf("%s-snssub", t.RunID)
	resp, err := g.sns().CreateTopic(ctx, &sns.CreateTopicInput{Name: aws.String(name)})
	if err != nil {
		return err
	}
	t.Set("sns_sub_topic_arn", aws.ToString(resp.TopicArn))

	qname := fmt.Sprintf("%s-snssubq", t.RunID)
	qresp, err := g.sqs().CreateQueue(ctx, &sqs.CreateQueueInput{QueueName: aws.String(qname)})
	if err != nil {
		return err
	}
	t.Set("sns_sub_q_url", aws.ToString(qresp.QueueUrl))

	attr, _ := g.sqs().GetQueueAttributes(ctx, &sqs.GetQueueAttributesInput{
		QueueUrl: qresp.QueueUrl, AttributeNames: []sqstypes.QueueAttributeName{"QueueArn"},
	})
	t.Set("sns_sub_q_arn", attr.Attributes["QueueArn"])
	return nil
}

func (g *snsGroup) teardownSubscriptions(ctx context.Context, t *harness.TestContext) error {
	if arn := t.GetString("sns_created_sub_arn"); arn != "" {
		g.sns().Unsubscribe(ctx, &sns.UnsubscribeInput{SubscriptionArn: aws.String(arn)}) //nolint:errcheck
	}
	if arn := t.GetString("sns_sub_topic_arn"); arn != "" {
		g.sns().DeleteTopic(ctx, &sns.DeleteTopicInput{TopicArn: aws.String(arn)}) //nolint:errcheck
	}
	if url := t.GetString("sns_sub_q_url"); url != "" {
		g.sqs().DeleteQueue(ctx, &sqs.DeleteQueueInput{QueueUrl: aws.String(url)}) //nolint:errcheck
	}
	return nil
}

func (g *snsGroup) Subscribe(ctx context.Context, t *harness.TestContext) error {
	topicARN := t.GetString("sns_sub_topic_arn")
	qARN := t.GetString("sns_sub_q_arn")
	resp, err := g.sns().Subscribe(ctx, &sns.SubscribeInput{
		TopicArn: aws.String(topicARN), Protocol: aws.String("sqs"), Endpoint: aws.String(qARN),
	})
	if err != nil {
		return err
	}
	t.Set("sns_created_sub_arn", aws.ToString(resp.SubscriptionArn))
	return nil
}

func (g *snsGroup) SubscribeSQS(ctx context.Context, t *harness.TestContext) error {
	topicARN := t.GetString("sns_sub_topic_arn")
	qARN := t.GetString("sns_sub_q_arn")
	resp, err := g.sns().Subscribe(ctx, &sns.SubscribeInput{
		TopicArn: aws.String(topicARN), Protocol: aws.String("sqs"), Endpoint: aws.String(qARN),
	})
	if err != nil {
		return err
	}
	t.Set("sns_created_sub_arn", aws.ToString(resp.SubscriptionArn))
	return nil
}

func (g *snsGroup) GetSubscriptionAttributes(ctx context.Context, t *harness.TestContext) error {
	subARN := t.GetString("sns_created_sub_arn")
	if subARN == "" {
		return fmt.Errorf("GetSubscriptionAttributes: no subscription ARN in context")
	}
	resp, err := g.sns().GetSubscriptionAttributes(ctx, &sns.GetSubscriptionAttributesInput{
		SubscriptionArn: aws.String(subARN),
	})
	if err != nil {
		return err
	}
	if resp.Attributes["Protocol"] != "sqs" {
		return fmt.Errorf("GetSubscriptionAttributes: expected Protocol=sqs, got %q", resp.Attributes["Protocol"])
	}
	return nil
}

func (g *snsGroup) SetSubscriptionAttributes(ctx context.Context, t *harness.TestContext) error {
	subARN := t.GetString("sns_created_sub_arn")
	if subARN == "" {
		return fmt.Errorf("SetSubscriptionAttributes: no subscription ARN in context")
	}
	_, err := g.sns().SetSubscriptionAttributes(ctx, &sns.SetSubscriptionAttributesInput{
		SubscriptionArn: aws.String(subARN),
		AttributeName:   aws.String("RawMessageDelivery"),
		AttributeValue:  aws.String("true"),
	})
	if err != nil {
		return err
	}
	resp, err := g.sns().GetSubscriptionAttributes(ctx, &sns.GetSubscriptionAttributesInput{
		SubscriptionArn: aws.String(subARN),
	})
	if err != nil {
		return err
	}
	if resp.Attributes["RawMessageDelivery"] != "true" {
		return fmt.Errorf("SetSubscriptionAttributes: RawMessageDelivery not set")
	}
	return nil
}

func (g *snsGroup) ListSubscriptions(ctx context.Context, t *harness.TestContext) error {
	resp, err := g.sns().ListSubscriptions(ctx, &sns.ListSubscriptionsInput{})
	if err != nil {
		return err
	}
	subARN := t.GetString("sns_created_sub_arn")
	if subARN == "" {
		return nil
	}
	for _, sub := range resp.Subscriptions {
		if aws.ToString(sub.SubscriptionArn) == subARN {
			return nil
		}
	}
	return fmt.Errorf("ListSubscriptions: subscription %q not found", subARN)
}

func (g *snsGroup) ListSubscriptionsByTopic(ctx context.Context, t *harness.TestContext) error {
	topicARN := t.GetString("sns_sub_topic_arn")
	resp, err := g.sns().ListSubscriptionsByTopic(ctx, &sns.ListSubscriptionsByTopicInput{
		TopicArn: aws.String(topicARN),
	})
	if err != nil {
		return err
	}
	subARN := t.GetString("sns_created_sub_arn")
	if subARN == "" {
		return nil
	}
	for _, sub := range resp.Subscriptions {
		if aws.ToString(sub.SubscriptionArn) == subARN {
			return nil
		}
	}
	return fmt.Errorf("ListSubscriptionsByTopic: subscription %q not found", subARN)
}

func (g *snsGroup) Unsubscribe(ctx context.Context, t *harness.TestContext) error {
	subARN := t.GetString("sns_created_sub_arn")
	if subARN == "" {
		return fmt.Errorf("Unsubscribe: no subscription ARN in context")
	}
	_, err := g.sns().Unsubscribe(ctx, &sns.UnsubscribeInput{SubscriptionArn: aws.String(subARN)})
	if err != nil {
		return err
	}
	t.Set("sns_created_sub_arn", "")
	// Verify subscription is gone
	topicARN := t.GetString("sns_sub_topic_arn")
	list, lErr := g.sns().ListSubscriptionsByTopic(ctx, &sns.ListSubscriptionsByTopicInput{
		TopicArn: aws.String(topicARN),
	})
	if lErr != nil {
		return nil // ListSubscriptionsByTopic may fail if topic was deleted
	}
	for _, sub := range list.Subscriptions {
		if aws.ToString(sub.SubscriptionArn) == subARN {
			return fmt.Errorf("Unsubscribe: subscription %q still present", subARN)
		}
	}
	return nil
}
